Goobi/src/de/sub/goobi/metadaten/Metadaten.java

```diff
@@ -3098,15 +3099,59 @@ boolean canCreate(MetadataGroupType type) {
      *             multi-select input
      */
     public List<RenderableMetadataGroup> getMyGroups() throws ConfigurationException {
-        List<MetadataGroup> records;
-        if (myDocStruct == null || (records = myDocStruct.getAllMetadataGroups()) == null) {
+        if (myDocStruct == null) {
             return Collections.emptyList();
         }
+
+        List<MetadataGroup> records = myDocStruct.getAllMetadataGroups();
+        if (records == null) {
+            records = Collections.emptyList();
+        }
+        List<MetadataGroupType> defaultGroupTypes = myDocStruct.getType().getAllDefaultDisplayMetadataGroups();
+        if (records.isEmpty() && defaultGroupTypes.isEmpty()) {
+            return Collections.emptyList();
+        }
+
         List<RenderableMetadataGroup> result = new ArrayList<RenderableMetadataGroup>(records.size());
         String language = (String) Helper.getManagedBeanValue("#{LoginForm.myBenutzer.metadatenSprache}");
         String projectName = myProzess.getProjekt().getTitel();
+        Map<String, List<MetadataGroup>> displayGroups = myProzess.getRegelsatz().isOrderMetadataByRuleset()
+                ? new LinkedHashMap<String, List<MetadataGroup>>() : new TreeMap<String, List<MetadataGroup>>();
+        for (MetadataGroupType groupType : myDocStruct.getType().getAllMetadataGroupTypes()) {
+            displayGroups.put(groupType.getName(), null);
+        }
+        for (MetadataGroupType groupType : defaultGroupTypes) {
+            displayGroups.put(groupType.getName(), new ArrayList());
+        }
         for (MetadataGroup record : records) {
-            result.add(new RenderableMetadataGroup(record, this, language, projectName));
+            String name = record.getType().getName();
+            if (displayGroups.get(name) == null) {
+                displayGroups.put(name, new ArrayList());
+            }
+            displayGroups.get(name).add(record);
+        }
+        for (Entry<String, List<MetadataGroup>> entry : displayGroups.entrySet()) {
+            records = entry.getValue();
+            if (records != null) {
+                if (records.size() > 0) {
+                    for (MetadataGroup record : records) {
+                        result.add(new RenderableMetadataGroup(record, this, language, projectName));
+                    }
+                } else {
+                    for (MetadataGroupType groupType : myDocStruct.getType().getAllMetadataGroupTypes()) {
+                        if (groupType.getName().equals(entry.getKey())) {
+                            RenderableMetadataGroup newMetadataGroup = new RenderableMetadataGroup(groupType, language, projectName);
+                            result.add(newMetadataGroup);
+                            try {
+                                myDocStruct.addMetadataGroup(newMetadataGroup.toMetadataGroup());
+                            } catch (DocStructHasNoTypeException | MetadataTypeNotAllowedException e) {
```

Comment:
why change exceptions here? either handle them here or hand them to the calling method

Reply:
I think that's the best way to treat them. Both exceptions cannot occur at this point. A DocStruct with no type cannot occur here, because the metadata group is defined by the type of DocStruct in the first place. The MetadataType cannot be not allowed either, otherwise the field would not have been previously generated by the code. The most correct treatment of these exceptions would be an empty catch-block, but I reject that in principle. That's why this treatment.